### PR TITLE
#68: Enforce TODO comment syntax

### LINKED ISSUE #68: Enforce TODO comment syntax
Ensure TODO comments are defined as follows in single line comments `//`:
- [x] Without developer name
```swift
// TODO: <insert mandatory todo comment>
```
- [x] With developer name
```swift
// TODO(dev-name): <insert mandatory todo comment>
```

**Note**: Nested single line comments will be analyzed and flagged.

### PR DESCRIPTION
**Pull request by adityatrivedi**
Resolves #68.

### CODE REVIEW

File: src/main/java/com/sleekbyte/tailor/listeners/TodoCommentListener.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void analyze() {
+        checkTodoSyntaxInSingleLineComments();
+        checkTodoSyntaxInMultiLineComments();
+    }
+
+    private void checkTodoSyntaxInSingleLineComments() {
+        String todoRegex = "(?s)// TODO: [^\\s].+";
```
Comment (a4sriniv) [inline src/main/java/com/sleekbyte/tailor/listeners/TodoCommentListener.java]: `// TODO(adithya): something goes here` is pretty common among developers and organizations (both Coursera and Google used this), so I think we should allow for that

File: src/main/antlr/Swift.g4 [hunk 2]
```diff
@@ -173,7 +173,7 @@ availabilityCondition: '#available' '(' availabilityArguments ')' ;
 availabilityArguments: availabilityArgument (',' availabilityArguments)? ;
 availabilityArgument: platformName platformVersion | '*' ;
 platformName: 'iOS' | 'iOSApplicationExtension' | 'OSX' | 'OSXApplicationExtension­' | 'watchOS' ;
-platformVersion: VersionLiteral | DecimalLiteral | FloatingPointLiteral ; // TODO: Find a way to make this only VersionLiteral
+platformVersion: VersionLiteral | DecimalLiteral | FloatingPointLiteral ; // TODO_SYNTAX: Find a way to make this only VersionLiteral
```
Comment (a4sriniv) [inline src/main/antlr/Swift.g4]: why did you do this?

Comment (adityatrivedi) [inline src/main/antlr/Swift.g4]: Oh lol, my bad.  It seems like a find/replace bug.  Will revert.

File: src/main/java/com/sleekbyte/tailor/listeners/TodoCommentListener.java (hunk 1 above)
Comment (adityatrivedi) [inline src/main/java/com/sleekbyte/tailor/listeners/TodoCommentListener.java]: Ok sounds good.

File: src/main/java/com/sleekbyte/tailor/Tailor.java [hunk 3]
```diff
@@ -141,9 +143,12 @@ public static void printSummary(long numFiles, long numSkipped, long numErrors,
 
                 if (className.equals(FileListener.class.getName())) {
                     continue;
-                } else if (className.equals(CommentAnalyzer.class.getName())) {
```
Comment (a4sriniv) [inline src/main/java/com/sleekbyte/tailor/Tailor.java]: maybe it's time to use a switch statement here .. I'll leave this up to your discretion

File: src/main/java/com/sleekbyte/tailor/common/Rules.java [hunk 4]
```diff
@@ -126,6 +127,10 @@ public String getLink() {
         TERMINATING_SEMICOLON.description = "Statements should not be terminated with semicolons.";
         TERMINATING_SEMICOLON.className = SemicolonTerminatedListener.class.getName();
 
+        TODO_SYNTAX.name = "todo-syntax";
+        TODO_SYNTAX.description = "Todo comments should follow <TODO_SYNTAX: description> format.";
```
Comment (a4sriniv) [inline src/main/java/com/sleekbyte/tailor/common/Rules.java]: why `TODO_SYNTAX` as opposed to `TODO`?

File: src/main/java/com/sleekbyte/tailor/listeners/CommentAnalyzer.java [hunk 5]
```diff
@@ -33,6 +30,8 @@ public CommentAnalyzer(CommonTokenStream tokenStream, Printer printer) {
         extractComments();
     }
 
+    public abstract void analyze();
+
     private void extractComments() {
         for (Token token : tokenStream.getTokens()) {
```
Comment (a4sriniv) [inline src/main/java/com/sleekbyte/tailor/listeners/CommentAnalyzer.java]: I think this process of extracting comments is expensive, so maybe we want a class dedicated to doing just that? We could extract the comments once and then we could just pass the list of comment tokens around among the different comment related listeners?

File: src/test/swift/com/sleekbyte/tailor/functional/TodoCommentTest.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    // this is an invalid comment because it contains TODO:todo Todo: words
+    /* this is also an invalid comment since it has TODO:todo:
+    Todo: words */
+
+    /* tOdo fix bug: return positive integer
+     */
+    func getPositiveNumber() -> Int {
+        // TODO: Valid single line todo comment
```
Comment (a4sriniv) [inline src/test/swift/com/sleekbyte/tailor/functional/TodoCommentTest.swift]: could you also add a test case with `TODO :`?

File: src/main/java/com/sleekbyte/tailor/common/Rules.java (hunk 4 above)
Comment (adityatrivedi) [inline src/main/java/com/sleekbyte/tailor/common/Rules.java]: I find it more explicit in communicating intent.

Comment (a4sriniv) [inline src/main/java/com/sleekbyte/tailor/common/Rules.java]: okay, we should probably elaborate on what TODO_SYNTAX can be in our docs

File: src/test/swift/com/sleekbyte/tailor/functional/TodoCommentTest.swift (hunk 6 above)
Comment (adityatrivedi) [inline src/test/swift/com/sleekbyte/tailor/functional/TodoCommentTest.swift]: Done

**Comment by a4sriniv** [comment]
:+1:

**Comment by a4sriniv** [comment]
:+1:

**Pull request merged**